src/mono/mono/mini/monovm.c

```diff
@@ -226,6 +226,40 @@ monovm_initialize (int propertyCount, const char **propertyKeys, const char **pr
 	return 0;
 }
 
+// Initialize monovm with properties set by runtimeconfig.json. Primarily used by mobile targets.
+int
+monovm_runtimeconfig_initialize (MonovmRuntimeConfigArguments *arg, MonovmRuntimeConfigArgumentsCleanup cleanup_fn, void *user_data)
+{
+	switch (arg->kind) {
+	case 0: {
+		MonoFileMap *map;
+		void *ret_handle;
+		char *buffer;
+
+		map = mono_file_map_open (arg->runtimeconfig.name.path);
+		g_assert (map);
+		arg->runtimeconfig.data.data_len = (uint32_t)mono_file_map_size (map);
+		g_assert (arg->runtimeconfig.data.data_len > 0);
+		buffer = (char*)mono_file_map (arg->runtimeconfig.data.data_len, MONO_MMAP_READ|MONO_MMAP_PRIVATE, mono_file_map_fd (map), 0, &ret_handle);
+		g_assert (buffer);
+
+		arg->runtimeconfig.data.data = g_new0 (char, arg->runtimeconfig.data.data_len);
+		memcpy (arg->runtimeconfig.data.data, buffer, sizeof(char) * arg->runtimeconfig.data.data_len);
+
+		mono_file_unmap (map, ret_handle);
+		mono_file_map_close (map);
```

Comment:
I would move this to `_install` as well.

Here's what I think install should look like:

```c

   /* ... utf8 to utf16 the appctx props ... */

   MonoFileMap *runtimeconfig_json_map = NULL;
   void* runtimeconfig_json_map_handle = NULL;
   const char *buffer_start = runtimeconfig_json_get_buffer (runtime_config_arg, &runtimeconfig_json_map, &runtimeconfig_json_map_handle);
   const char *buffer = buffer_start;

   int runtimeconfig_json_nprops = mono_metadata_decode_value (buffer, &buffer);
   
   int n_total_props = n_appctx_props + runtimeconfig_json_nprops;

   gunchar2** total_keys = g_new0 (gunichar2*, n_total_props);
   gunchar2** total_values = g_new0 (gunichar2*, n_total_props);
  
  /* ... populate the appctx props ...*/
 
  runtimeconfig_json_read_props (buffer, &buffer, runtimeconfig_json_nprops, total_keys+n_appctx_props, total_values+n_appctx_props);

  if (map != NULL) {
    mono_file_unmap (buffer_start, runtimeconfig_json_map_handle);
    mono_file_map_close (runtimeconfig_json_map);
  }

  if (runtime_config_cleanup_fn)
    runtime_config_cleanup_fn (runtime_config_arg, runtime_config_user_data);

  /* ... do the invoke ... */

  /* ... free the total_props ... */
```

The helper functions would be:

```c
/* if the runtime config arg is data, jut return the buffer and write NULL to *file_map and *buf_handle.
 * if the runtime config arg is a filename, mmap the file, and set *file_map to the opened file map, and *buf_pointer to the file map and the handle.
 */
static const char*
runtimeconfig_json_get_buffer (MonovmRuntimeConfigArguments *arg, MonoFileMap** file_map, gpointer * buf_handle);

/* start reading (length-prefixed) keys&values from ptr and writing their utf16 versions to dest_keys[0], dest_keys[1]... and dest_values[0], dest_values[1], ....   reading nprops pairs total.  at the end write the pointer location to *endp
 */
static void
runtimeconfig_json_read_props (const char *ptr, const char **endp, int nprops, guinchar2 **dest_keys, gunichar2 **dest_values);
```

Reply:
Ah, let me implement it this way.